### PR TITLE
Provenance v1: raise prominence of extensions + other cleanups

### PR DESCRIPTION
**Pull request by MarkLodato**
- Fix type of `builder.version`
- Sort fields of Builder alphabetically
- Better document extension fields so that it is more prominent
- Remove stale RFC tags
- Update changelog to reflect latest changes.

The main change is to raise the prominence of extension fields so that implementers know that it's an option.

### CODE REVIEW

File: docs/provenance/v1.md [hunk 1]
```diff
@@ -361,6 +348,11 @@ Dependencies used by the orchestrator that are not run within the workload
 and that do not affect the build, but might affect the provenance generation
 or security guarantees.
 
+<tr id="builder.version"><td><code>version</code>
+<td>map (string→string)<td>
+
+Version numbers of components of the build platform.
```
Comment (laurentsimon) [inline docs/provenance/v1.md:354]: Thoughts on re-using `ArtifactDescriptor` which can record richer information? Or is `version` the only relevant information for the platform components / dependencies? 
Would it a bad idea to add a version field to `ArtifactDescriptor`?

Comment (TomHennen) [reply docs/provenance/v1.md:354]: FWIW I'm also curious about this field, it's not clear what the key is supposed to be?

Replacing `version` with some sort of ResourceDescriptor list (as Laurent is suggesting) might be more interesting?  But also would result in less information actually in the provenance or be pretty awkward to express inline.

I could also see how any change like that would be better off in a different PR (since this is just alphabetizing)?

**Review by TomHennen — APPROVED** [review]

File: docs/provenance/v1.md (hunk 1 above)
Comment (MarkLodato) [reply docs/provenance/v1.md:354]: The intention is that `version` is just for string values. To record the actual artifacts used by the builder, there is `builderDependencies`.

I think such a change is outside the scope of this PR.

File: docs/provenance/v1.md [hunk 2]
```diff
@@ -433,6 +418,26 @@ The timestamp of when the build completed.
 
 </table>
 
+### Extension fields
+
+[Extension fields]: #extension-fields
```
Comment (laurentsimon) [inline docs/provenance/v1.md:423]: so these fields are not in an "extension" namespace?

File: docs/provenance/v1.md [hunk 3]
```diff
@@ -433,6 +418,26 @@ The timestamp of when the build completed.
...
+
+Implementations MAY add extension fields to any JSON object to describe
```
Comment (laurentsimon) [inline docs/provenance/v1.md:425]: What happens if the next version uses a field that is already used by someone else's extension?

Comment (MarkLodato) [reply docs/provenance/v1.md:425]: It will be ambiguous what the author meant. Are you requesting a change in the doc?

Comment (laurentsimon) [reply docs/provenance/v1.md:425]: That could help, but I suspect extension are buildType dependent so it should be OK?

File: docs/provenance/v1.md (hunk 2 above)
Comment (laurentsimon) [reply docs/provenance/v1.md:423]: Yes, I find it hard to avoid collisions, in particular if more future versions of SLSA specs may add new fields. That makes it impossible to define extension names that don't collide with future versions. If we at least say that the extension should start with `ext_`, then this would be future-proof. The namespace is another way to do it.

File: docs/provenance/v1.md (hunk 3 above)
Comment (laurentsimon) [reply docs/provenance/v1.md:425]: But also, telling people how to name extensions to avoid collision could be useful. Ideally they would be unique to builderID+buildType, e.g. `builderID_BuildType_NAME` although that would make the fields rather long :(

Comment (TomHennen) [reply docs/provenance/v1.md:425]: FWIW I'm imagining doing something like `<product|company name>_fieldName`.  E.g. `gerrit_codeReview`

Comment (lumjjb) [reply docs/provenance/v1.md:425]: In the same vein, i also do like what OCI does with using the domains like "com.google.xyz" that way

Comment (MarkLodato) [reply docs/provenance/v1.md:425]: PTAL at 3527436, which uses Tom's suggestion of `<vendor>_<name>`. I think that's a good compromise between full collision avoidance and having a concise, easy-to-read name.

I made it a MUST but I'm not confident about that.

Note: We originally suggested using full URLs as the field name to avoid collisions, but that's very verbose and also hard to parse due to the `.`s. A reverse domain name seems unnecessarily complicated to me.

File: docs/provenance/v1.md (hunk 1 above)
Comment (joshuagl) [inline docs/provenance/v1.md:354]: This is a map of component->version, right? Can we somehow clarify that in the description?
```suggestion
Map of names of components of the build platform to their version.
```

**Review by joshuagl — APPROVED** [review]
I like the change to prefix extensions. LGTM, thanks!

File: docs/provenance/v1.md (hunk 3 above)
Comment (laurentsimon) [reply docs/provenance/v1.md:427]: LGTM.

File: docs/provenance/v1.md [hunk 4]
```diff
@@ -435,6 +420,25 @@ The timestamp of when the build completed.
...
+information that is not captured in a standard field. Guidelines:
+
+-   Extension fields MUST use names of the form `<vendor>_<fieldname>`, e.g.
+    `examplebuilder_isCodeReviewed`. This practice avoids field name collisions
+    by namespacing each vendor. Non-extension field names never contain an
+    underscore.
```
Comment (lehors) [inline docs/provenance/v1.md:433]: While this is a reasonable way to provide for what we cannot anticipate it should be noted that this kind of extension mechanism has some inherent flaws. Namely, experience shows that it becomes very difficult to ever change those names. For an example, consider all the x-* HTTP headers. See:

> Custom proprietary headers have historically been used with an X- prefix, but this convention was deprecated in June 2012 because of the inconveniences it caused when nonstandard fields became standard

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers

Comment (MarkLodato) [reply docs/provenance/v1.md]: I downgraded this to a SHOULD to give us flexibility in the future, but I do think the convention is useful. Consensus seems to be that we'd prefer to avoid naming collisions, even at the expense of uglier names. In the future if we find this to be a problem, we can always come up with a better system. But the bigger problem seems like naming collisions now than ugly names in the future.

**Review by lehors — APPROVED** [review]
